arda/arda.py

```diff
@@ -0,0 +1,79 @@
+import numpy as np
+from sklearn.ensemble import RandomForestRegressor
+
+
+# algo 2
+# A                     the (normalized) data matrix
+# n                     the amount of features to generate
+#
+# Return: A matrix of generated random features, where each column represents one feature
+def gen_features(A, n):
+    L = []
+    m = np.mean(A, axis=1)
+    s = np.cov(A)
+    for i in range(n):
+        L.append(np.random.multivariate_normal(m, s))
+    return np.array(L).T
+
+
+# algo 1
+# A                     the (normalized) data matrix
+# y                     the feature to use as criterion/dependent variable for the regressors
+# tau                   threshold for how many times a feature appeared in front of synthesized features in ranking
+# eta                   fraction of random features to inject (fraction of amount of features in A)
+# k                     number of times ranking and counting is performed
+# regressor             the regressor to use
+#
+# Returns: An array of indices, corresponding to selected features from A
+def select_features(A, y, tau=0.1, eta=0.5, k=20, regressor=RandomForestRegressor):
+    d = A.shape[1]
+    augment_count = round(eta * d)
+    X = np.concatenate((A, gen_features(A, augment_count)), axis=1)  # This gives us A' from the paper
+    agg = np.zeros(d)
+
+    # Repeat process 'k' times, as in the algorithm
+    for i in range(k):
+        reg = regressor()
+        mask = np.zeros(d + augment_count, dtype=bool)
+        mask[d:] = True  # We mark the columns that were generated
+        reg.fit(X, y)
+        indices = reg.feature_importances_.argsort()[::-1]  # Then we obtain sorting indices for the rankings
+        sorted_mask = mask[indices[::]]  # These indices are then used to sort the mask, so we know where the generated columns are located in terms of ranking
+        
+        # Then we iterate through this mask until we hit a generated feature, adding 1 for all the original features that were in front
+        for i, val in zip(indices, sorted_mask):
```

Comment:
Use list comprehension instead.

Reply:
Can you clarify? The loop here is not used to build a list, instead it just finds the first random column and then stops.